# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [4]:
# Libraries
import scipy as sp
import numpy as np
import pandas as pd
import datetime
import sklearn
from sklearn import datasets
pd.set_option('display.max_columns', None)

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [17]:
# your code here
diabetes=datasets.load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [43]:
# your code here
diabetes.keys()

dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [20]:
# your code here
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, T-Cells (a type of white blood cells)
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, thyroid stimulating hormone
      - s5      ltg, lamotrigine
      - s6      glu, blood sugar level

Note: Each of these 10 feature va

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

### your answer here

1. There are 10 attributes in the dataset. They show different medical metrics of people with diabetes.
2. The data is presumably causally related to the target variable, which measures the disease progression one year after the base time.
3. There are only 442 records.

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [44]:
# your code here
df=pd.DataFrame(diabetes["data"])
target=pd.DataFrame(diabetes["target"])

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:


$y = \beta_0 + \beta_1·X_1 + \beta_2·X_2 + ... + \beta_n·X_n + \epsilon$


where:

- $X_1-X_n$: data 
- $\beta_0$: intercept 
- $\beta_1-\beta_n$: coefficients 
- $\epsilon$: error (cannot explained by model)
- $y$: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [49]:
# your code here
from sklearn.linear_model import LinearRegression

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [57]:
# your code here
reg=LinearRegression()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [56]:
# your code here
diabetes_data_train=df.head(len(df)-20)
diabetes_data_test=df.tail(20)
diabetes_target_train=target.head(len(df)-20)
diabetes_target_test=target.tail(20)

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [107]:
# your code here
reg.fit(diabetes_data_train,diabetes_target_train)

print(reg.score(diabetes_data_train,diabetes_target_train))

print(reg.intercept_)

print(reg.coef_)

0.5122262147793499
[152.76430692]
[[ 3.03499549e-01 -2.37639315e+02  5.10530605e+02  3.27736980e+02
  -8.14131709e+02  4.92814588e+02  1.02848452e+02  1.84606489e+02
   7.43519617e+02  7.60951722e+01]]


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [106]:
# your code here
print(reg.predict(diabetes_data_test))
print(diabetes_target_test)

[[197.61846908]
 [155.43979328]
 [172.88665147]
 [111.53537279]
 [164.80054784]
 [131.06954875]
 [259.12237761]
 [100.47935157]
 [117.0601052 ]
 [124.30503555]
 [218.36632793]
 [ 61.19831284]
 [132.25046751]
 [120.3332925 ]
 [ 52.54458691]
 [194.03798088]
 [102.57139702]
 [123.56604987]
 [211.0346317 ]
 [ 52.60335674]]
         0
422  233.0
423   91.0
424  111.0
425  152.0
426  120.0
427   67.0
428  310.0
429   94.0
430  183.0
431   66.0
432  173.0
433   72.0
434   49.0
435   64.0
436   48.0
437  178.0
438  104.0
439  132.0
440  220.0
441   57.0


#### Print your `diabetes_target_test` and compare with the prediction. 

In [102]:
# your code here

np.std((np.array(diabetes_target_test)-reg.predict(diabetes_data_test))/(np.array(diabetes_target_test))
)

0.5141294316080605

In [105]:
y_test_pred = reg.predict(diabetes_data_test)
sklearn.metrics.r2_score(diabetes_target_test, y_test_pred)

0.5850753022690571

#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

### Your answer here
No. They differ considerably, and R-squared is only .6

# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://online.stat.psu.edu/stat501/lesson/6/6.4) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    $H_0 : \beta_1 = \beta_2 = ... = \beta_{10} = 0$

    $H_a$ : At least one $\beta_j \ne 0$ (for j = 1, 2, ..., 10)
    
    
* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

<div style="text-align:center"><img src="../images/statsmodels.png" width=600/></div>

In [111]:
# your code here



#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?

In [ ]:
# your answer here

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [145]:
# your code here
auto=pd.read_csv(('../data/auto-mpg.csv'))

Look at the first 5 rows using the `head()` function:

In [146]:
# your code here
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [147]:
# your code here
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

What is the newest model year and the oldest model year?

In [148]:
# your code here
print(auto["model_year"].min())
print(auto["model_year"].max())

70
82


Check the dataset for missing values and remove all rows containing at least one missing value.

In [158]:
# your code here
auto.dropna(axis=0,inplace=True)
auto.reset_index(drop=True)

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0    18.0          8         307.0        130.0    3504          12.0   
1    15.0          8         350.0        165.0    3693          11.5   
2    18.0          8         318.0        150.0    3436          11.0   
3    16.0          8         304.0        150.0    3433          12.0   
4    17.0          8         302.0        140.0    3449          10.5   
..    ...        ...           ...          ...     ...           ...   
387  27.0          4         140.0         86.0    2790          15.6   
388  44.0          4          97.0         52.0    2130          24.6   
389  32.0          4         135.0         84.0    2295          11.6   
390  28.0          4         120.0         79.0    2625          18.6   
391  31.0          4         119.0         82.0    2720          19.4   

     model_year                       car_name  
0            70  \t"chevrolet chevelle malibu"  
1            70          \t"buick skylark 320"  
2            70         \t"plymouth satellite"  
3            70              \t"amc rebel sst"  
4            70                \t"ford torino"  
..          ...                            ...  
387          82            \t"ford mustang gl"  
388          82                  \t"vw pickup"  
389          82              \t"dodge rampage"  
390          82                \t"ford ranger"  
391          82                 \t"chevy s-10"  

[392 rows x 8 columns]

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [166]:
auto["cylinders"].value_counts()
#There are 5 cylinder types

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [ ]:
auto.drop("car_name",axis="columns",inplace=True)

In [175]:
# your code here
from sklearn.model_selection import train_test_split

y=auto["mpg"]
x=auto.drop("mpg",axis="columns")

x_train, x_test, y_train, y_test=train_test_split(x,y,test_size=.2)
print(y_train.shape,y_test.shape)
x_train.shape,x_test.shape

(313,) (79,)


((313, 6), (79, 6))

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [179]:
# your code here
automodel=LinearRegression()
automodel.fit(x_train,y_train)

LinearRegression()


# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](https://en.wikipedia.org/wiki/Coefficient_of_determination). Its formula is:

<div style="text-align:center"><img src="../images/r-squared.png" width=300/></div>

where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [187]:
# your code here
from sklearn import metrics


y_train_pred = automodel.predict(x_train)
metrics.r2_score(y_train,y_train_pred)

0.818620364779308

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [186]:
# your code here


y_test_pred = automodel.predict(x_test)
metrics.r2_score(y_test,y_test_pred)

0.7574580409003677

#### Explaining the results

Please, compare the results for the training and test sets and comment. What can you say?

### your answer here

The model obviously performs better on the training data with an R-squared of 0.82. Nevertheless, the score for the test data is not bad at 0.76. This means that the predictor variables account for 76% of the variation in y on the test set.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [188]:
# your code here

x_train09, x_test09, y_train09, y_test09=train_test_split(x,y,test_size=.1)
print(y_train.shape,y_test.shape)
x_train.shape,x_test.shape

(313,) (79,)


((313, 6), (79, 6))

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [199]:
# your code here
auto_model09=LinearRegression()

Compute the predicted values and r squared score for our new model and new sample data.

In [206]:
# your code here

y_train_pred = auto_model09.predict(x_train09)

metrics.r2_score(y_train09,y_train_pred)

auto_model09.fit(x_train09,y_train09)
auto_model09.score(x_train09,y_train09)

0.8055695790759498

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [207]:
# your code here

y_test_pred = auto_model09.predict(x_test09)
metrics.r2_score(y_test09,y_test_pred)

0.8363244830235115

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE.

In [ ]:
# Libraries
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`.

In [214]:
# your code here

from sklearn.feature_selection import RFE

selector=RFE(automodel,n_features_to_select=3)

Fit the model and print the ranking.

In [215]:
# your code here
selector=selector.fit(x_train,y_train)

In [220]:
selector.ranking_

array([1, 4, 3, 2, 1, 1])

Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [221]:
auto

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0    18.0          8         307.0        130.0    3504          12.0   
1    15.0          8         350.0        165.0    3693          11.5   
2    18.0          8         318.0        150.0    3436          11.0   
3    16.0          8         304.0        150.0    3433          12.0   
4    17.0          8         302.0        140.0    3449          10.5   
..    ...        ...           ...          ...     ...           ...   
393  27.0          4         140.0         86.0    2790          15.6   
394  44.0          4          97.0         52.0    2130          24.6   
395  32.0          4         135.0         84.0    2295          11.6   
396  28.0          4         120.0         79.0    2625          18.6   
397  31.0          4         119.0         82.0    2720          19.4   

     model_year  
0            70  
1            70  
2            70  
3            70  
4            70  
..          ...  
393          82  
394          82  
395          82  
396          82  
397          82  

[392 rows x 7 columns]

In [223]:
# your code here
# your code here
from sklearn.model_selection import train_test_split

y_rfe=auto["mpg"]
x_rfe=auto.drop(["mpg","cylinders","acceleration","model_year"],axis="columns")

x_train_rfe, x_test_rfe, y_train_rfe, y_test_rfe=train_test_split(x_rfe,y_rfe,test_size=.2)

Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [228]:
# your code here

# your code here
auto_model_rfe=LinearRegression()

auto_model_rfe.fit(x_train_rfe,y_train_rfe)


y_train_pred_rfe = auto_model_rfe.predict(x_train_rfe)


metrics.r2_score(y_train_rfe,y_train_pred_rfe)

auto_model_rfe.fit(x_train_rfe,y_train_rfe)
auto_model_rfe.score(x_train_rfe,y_train_rfe)

0.6938154807181884

#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.